arch: all: Remove legacy code

ceolin opened the pull request:
The struct _kernel_ach exists only because ARC' s port needed it, in
all other ports this was defined as an empty struct. Turns out that
this struct is not required even for ARC anymore, this is a legacy
code from nanokernel time.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
arch: all: Remove legacy code

The struct _kernel_ach exists only because ARC' s port needed it, in
all other ports this was defined as an empty struct. Turns out that
this struct is not required even for ARC anymore, this is a legacy
code from nanokernel time.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

tyrantlizard77 reviewed (changes requested):
I'd rather this abstraction was not removed, I'd planned on using it for some upcoming memory protection work.

If you want to get rid of the empty structs, you could have arch/cpu.h define Z_ARCH_HAS_KERNEL_ARCH or something and ifdef the definition of the arch member based on it.

ceolin commented:
> I'd rather this abstraction was not removed, I'd planned on using it for some upcoming memory protection work.
> 
> If you want to get rid of the empty structs, you could have arch/cpu.h define Z_ARCH_HAS_KERNEL_ARCH or something and ifdef the definition of the arch member based on it.

without future changes this is just legacy and not used code :/ We can keep it for future work, but at the time this patch was suggested it makes sense IMHO.
So, what you suggest, you want me to keep it what condition ? If I add this guard any platform will need it currently

Commit:
arch: all: Remove not used struct _caller_saved

The struct _caller_saved is not used. Most architectures put
automatically the registers onto stack, in others architectures the
exception code does it.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

andrewboie commented:
> > I'd rather this abstraction was not removed, I'd planned on using it for some upcoming memory protection work.
> > If you want to get rid of the empty structs, you could have arch/cpu.h define Z_ARCH_HAS_KERNEL_ARCH or something and ifdef the definition of the arch member based on it.
> 
> without future changes this is just legacy and not used code :/ We can keep it for future work, but at the time this patch was suggested it makes sense IMHO.
> So, what you suggest, you want me to keep it what condition ? If I add this guard any platform will need it currently

No you won't.

```
diff --git a/kernel/include/kernel_structs.h b/kernel/include/kernel_structs.h
index dd625d5844..909efe5844 100644
--- a/kernel/include/kernel_structs.h
+++ b/kernel/include/kernel_structs.h
@@ -168,8 +168,10 @@ struct z_kernel {
        struct k_thread *threads; /* singly linked list of ALL threads */
 #endif
 
+#ifdef Z_ARCH_HAS_KERNEL_ARCH_MEMBER
        /* arch-specific part of _kernel */
        struct _kernel_arch arch;
+#endif
 };
 
 typedef struct z_kernel _kernel_t;
```

You are done and can delete all the empty struct _kernel_arch defintions in the arch headers

ceolin commented:
> > > I'd rather this abstraction was not removed, I'd planned on using it for some upcoming memory protection work.
> > > If you want to get rid of the empty structs, you could have arch/cpu.h define Z_ARCH_HAS_KERNEL_ARCH or something and ifdef the definition of the arch member based on it.
> > 
> > 
> > without future changes this is just legacy and not used code :/ We can keep it for future work, but at the time this patch was suggested it makes sense IMHO.
> > So, what you suggest, you want me to keep it what condition ? If I add this guard any platform will need it currently
> 
> No you won't.
> 
> ```
> diff --git a/kernel/include/kernel_structs.h b/kernel/include/kernel_structs.h
> index dd625d5844..909efe5844 100644
> --- a/kernel/include/kernel_structs.h
> +++ b/kernel/include/kernel_structs.h
> @@ -168,8 +168,10 @@ struct z_kernel {
>         struct k_thread *threads; /* singly linked list of ALL threads */
>  #endif
>  
> +#ifdef Z_ARCH_HAS_KERNEL_ARCH_MEMBER
>         /* arch-specific part of _kernel */
>         struct _kernel_arch arch;
> +#endif
>  };
>  
>  typedef struct z_kernel _kernel_t;
> ```
> 
> You are done and can delete all the empty struct _kernel_arch defintions in the arch headers

yeah, but my commit will still remove `struct _kernel_arch` definition, since they are being defined empty. Shouldn't be better you add it when you add code that needs it ? Other way is remove this commit by now

ceolin commented:
> > > I'd rather this abstraction was not removed, I'd planned on using it for some upcoming memory protection work.
> > > If you want to get rid of the empty structs, you could have arch/cpu.h define Z_ARCH_HAS_KERNEL_ARCH or something and ifdef the definition of the arch member based on it.
> > 
> > 
> > without future changes this is just legacy and not used code :/ We can keep it for future work, but at the time this patch was suggested it makes sense IMHO.
> > So, what you suggest, you want me to keep it what condition ? If I add this guard any platform will need it currently
> 
> No you won't.
> 
> ```
> diff --git a/kernel/include/kernel_structs.h b/kernel/include/kernel_structs.h
> index dd625d5844..909efe5844 100644
> --- a/kernel/include/kernel_structs.h
> +++ b/kernel/include/kernel_structs.h
> @@ -168,8 +168,10 @@ struct z_kernel {
>         struct k_thread *threads; /* singly linked list of ALL threads */
>  #endif
>  
> +#ifdef Z_ARCH_HAS_KERNEL_ARCH_MEMBER
>         /* arch-specific part of _kernel */
>         struct _kernel_arch arch;
> +#endif
>  };
>  
>  typedef struct z_kernel _kernel_t;
> ```
> 
> You are done and can delete all the empty struct _kernel_arch defintions in the arch headers

ok, this means that I'll remove this struct for *all* architectures. This piece of code won't make sense because any port will satisfy this condition right now. Won't be better have it when you use it ?

andyross reviewed (approved):
I'm actually with @ceolin here, I don't see much value to this struct as it stands.  The is and will only ever be one _kernel struct.  The reason other architectures don't have the struct defined isn't that they don't need their own global data, it's because they put it in their own global storage (static variables, yada yada).  I don't see why new work can't continue to do this.  In fact the kernel as a whole is really inconsistent on this.  Some stuff is a field in _kernel and some stuff isn't, and there's no rhyme or reason.

That said the **motiviation** for this change (that empty structs are somehow "bad") is IMHO ridiculous.  But I think it's useful cleanup regardless.

ceolin commented:
> I'm actually with @ceolin here, I don't see much value to this struct as it stands. The is and will only ever be one _kernel struct. The reason other architectures don't have the struct defined isn't that they don't need their own global data, it's because they put it in their own global storage (static variables, yada yada). I don't see why new work can't continue to do this. In fact the kernel as a whole is really inconsistent on this. Some stuff is a field in _kernel and some stuff isn't, and there's no rhyme or reason.
> 
> That said the **motiviation** for this change (that empty structs are somehow "bad") is IMHO ridiculous. But I think it's useful cleanup regardless.

The motivation is not that, it is the fact it is not part o C standard, this is a compiler extension and make us less portable. These cases were really nice actually, it was an opportunity to remove things not used from kernel, it did not require any additional change.
